### PR TITLE
[CLI-2569] Check for magic byte when consuming avro/protobuf/jsonschema messages

### PR DESCRIPTION
**Pull request by sgagniere**
Release Notes
-------------

Breaking Changes
- PLACEHOLDER

New Features
- PLACEHOLDER

Bug Fixes
- Fix misleading error messages in `confluent kafka topic consume`

Checklist
---------
1. [CRUCIAL] Is the change for features that are already live in prod?
   * yes: ok

What
----
Consuming a message which doesn't follow the correct format (e.g. running `confluent kafka topic consume --value-format avro` and then attempting to consume the string `abcd`) currently causes one of two errors, both of which are misleading:
- `Error: failed to find schema ID in topic data` if the message is short, or
- `Error: {"error_code":40301, "message":"User is denied operation on this server."}` if the message is long enough.

This PR adds an additional check that the magic byte has the correct value when `--value-format` is not set to `string`, and returns the same error a user would see if they were using Kafka console consumer: `unknown magic byte`.

References
----------

Test & Review
-------------
Manual testing (since consume processes messages in a somewhat random order, I reran these commands a few times to get the correct and incorrect messages to both show in the outputs below)
Avro success:
```
confluent kafka topic produce avro-test --value-format avro --schema ~/schemas/schema-avro-test-value-v1.avsc --schema-registry-api-key $APIKEY --schema-registry-api-secret $APISECRET
Successfully registered schema with ID 100003
Starting Kafka Producer. Use Ctrl-C or Ctrl-D to exit.
{"my_field1":1, "my_field2":1.0, "my_field3":""}
```
```
confluent kafka topic consume avro-test -b --value-format avro --schema-registry-api-key $APIKEY --schema-registry-api-secret $APISECRET                
Starting Kafka Consumer. Use Ctrl-C to exit.
{"my_field2":1,"my_field3":"","my_field1":1}
```
Avro fail:
```
confluent kafka topic produce avro-test
Starting Kafka Producer. Use Ctrl-C or Ctrl-D to exit.
abcd
```
```
confluent kafka topic consume avro-test -b --value-format avro --schema-registry-api-key $APIKEY --schema-registry-api-secret $APISECRET
Starting Kafka Consumer. Use Ctrl-C to exit.
{"my_field1":1,"my_field2":1,"my_field3":""}
Error: unknown magic byte
```

Protobuf success:
```
make && confluent kafka topic produce protobuf-test --value-format protobuf --schema ~/schemas/schema-protobuf-test-value-v1.proto --schema-registry-api-key $APIKEY --schema-registry-api-secret $APISECRET
Successfully registered schema with ID 100009
Starting Kafka Producer. Use Ctrl-C or Ctrl-D to exit.
{"my_field1":1, "my_field2":1.0, "my_field3":"one"}
```
```
confluent kafka topic consume protobuf-test -b --value-format protobuf --schema-registry-api-key $APIKEY --schema-registry-api-secret $APISECRET
Starting Kafka Consumer. Use Ctrl-C to exit.
{"myField1":1,"myField2":1,"myField3":"one"}
```
Protobuf fail:
```
confluent kafka topic produce protobuf-test
Starting Kafka Producer. Use Ctrl-C or Ctrl-D to exit.
asdf
```
```
confluent kafka topic consume protobuf-test -b --value-format protobuf --schema-registry-api-key $APIKEY --schema-registry-api-secret $APISECRET
Starting Kafka Consumer. Use Ctrl-C to exit.
{"myField1":1,"myField2":1,"myField3":"one"}
Error: unknown magic byte
```

Jsonschema success:
```
confluent kafka topic produce json-test --value-format jsonschema --schema ~/schemas/schema-json-test-value-v1.json --schema-registry-api-key $APIKEY --schema-registry-api-secret $APISECRET
Successfully registered schema with ID 100011
Starting Kafka Producer. Use Ctrl-C or Ctrl-D to exit.
{"myField1":1,"myField2":1,"myField3":"one"}
```
```
confluent kafka topic consume json-test -b --value-format jsonschema --schema-registry-api-key $APIKEY --schema-registry-api-secret $APISECRET   
Starting Kafka Consumer. Use Ctrl-C to exit.
{"myField1":1,"myField2":1,"myField3":"one"}
```
Jsonschema fail:
```
confluent kafka topic produce json-test
Starting Kafka Producer. Use Ctrl-C or Ctrl-D to exit.
abcd
```
```
confluent kafka topic consume json-test -b --value-format jsonschema --schema-registry-api-key $APIKEY --schema-registry-api-secret $APISECRET                                               
Starting Kafka Consumer. Use Ctrl-C to exit.
{"myField1":1,"myField2":1,"myField3":"one"}
Error: unknown magic byte
```

As a comparison, here's what currently happens:
```
confluent kafka topic consume json-test -b --value-format jsonschema --schema-registry-api-key $APIKEY --schema-registry-api-secret $APISECRET
Starting Kafka Consumer. Use Ctrl-C to exit.
{"myField1":1,"myField2":1,"myField3":"one"}
Error: failed to find schema ID in topic data
```
```
confluent kafka topic produce json-test                                                                                                       
Starting Kafka Producer. Use Ctrl-C or Ctrl-D to exit.
1234567890
```
```
confluent kafka topic consume json-test -b --value-format jsonschema --schema-registry-api-key $APIKEY --schema-registry-api-secret $APISECRET
Starting Kafka Consumer. Use Ctrl-C to exit.
Error: {"error_code":40301,"message":"User is denied operation on this server."}
```
Open Questions / Follow-ups
---------------------------

### CODE REVIEW

File: internal/cmd/kafka/confluent_kafka.go [hunk 1]
```diff
@@ -274,8 +274,11 @@ func RunConsumer(consumer *ckafka.Consumer, groupHandler *GroupHandler) error {
 }
 
 func (h *GroupHandler) RequestSchema(value []byte) (string, map[string]string, error) {
+	if len(value) == 0 || value[0] != 0x0 {
+		return "", nil, errors.New("unknown magic byte")
```
Comment (sgagniere) [inline internal/cmd/kafka/confluent_kafka.go:278]: Should I add a suggestion here? Maybe something about incorrect message formatting?

**Comment by MuweiHe** [comment]
is there always a magic byte? feel like plaintext messages don't have one.

**Comment by sgagniere** [comment]
@MuweiHe No, plaintext messages don't have a magic byte. This function `RequestSchema` is only called in two places, and both appear to check that the format isn't `string`.

**Review by MuweiHe — APPROVED** [review]

**Comment by brianstrauch** [comment]
> Resolve an issue in confluent kafka topic consume causing misleading error messages

How about "Fix misleading error messages in `confluent kafka topic consume`"?

**Comment by brianstrauch** [comment]
Would it help to say what the unknown magic byte is?

> unknown magic byte "\x00"

File: internal/cmd/kafka/confluent_kafka.go (hunk 1 above)
Comment (brianstrauch) [reply internal/cmd/kafka/confluent_kafka.go:278]: Yes! I'm not sure what "unknown magic byte" means to be honest

Comment (sgagniere) [reply internal/cmd/kafka/confluent_kafka.go]: It's a value used to validate the file format.

**Review by brianstrauch — APPROVED** [review]
+1, pending an answer to my last question

**Comment by sgagniere** [comment]
@brianstrauch I don't think it would be very helpful to write the magic byte that was found. With that information they could in principle look up what the file format is using magic byte tables online, but the more common scenario is that they produced a plaintext string, in which case the magic byte itself could be random.

**Pull request merged**